assert_series_equal considers Nan is not different from real value

### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.


### Issue description

assert_series_equal considers both Nan and real value as same.

I think this is because `testing.asserts._assert_series_inner` only check both left and right are null or not.
https://github.com/pola-rs/polars/blob/b311405b80b2a086d58450c64e7e7c2d2cbcb51d/py-polars/polars/testing/asserts.py#L188

For example, let left is a series filled with nan values, and right is a series filled with a real values. 
Then `left.is_null()` returns a series filled with False, and `right.is_null()` returns a series filled with False. 
So both series are considered as same when checking missing values.

### Suggestion
So, how about change https://github.com/pola-rs/polars/blob/b311405b80b2a086d58450c64e7e7c2d2cbcb51d/py-polars/polars/testing/asserts.py#L188 like as follows?

```
(left.fill_nan(None).is_null() != right.fill_nan(None).is_null()).any() 
```

If it's okay, can I work on it and submit a pull request?

### Reproducible example

```python
import polars as pl
from polars.testing import assert_series_equal

arr1 = pl.Series("a", [1.0, 2.0, 3.0, 4.0, 5.0])
arr2 = pl.Series("b", [1.0, 2.0, 3.0, 4.0, float("nan")])
arr3 = pl.Series("c", [1.0, 2.0, 3.0, 4.0, None])

assert_series_equal(arr1, arr2, check_names=False) # case 1
assert_series_equal(arr1, arr3, check_names=False) # case 2
```


### Expected behavior

both case 1 and case 2 must raise `AssertionError: Series are different`
however, only case 2 raise AssertionError, whereas case 1 is considered as equal series

### Installed versions

<details>

```
---Version info---
Polars: 0.15.2
Index type: UInt32
Platform: Linux-5.15.74.2-microsoft-standard-WSL2-x86_64-with-glibc2.31
Python: 3.9.12 (main, Apr 26 2022, 13:27:20) 
[GCC 9.4.0]
---Optional dependencies---
pyarrow: <not installed>
pandas: 1.5.2
numpy: 1.22.4
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: <not installed>
matplotlib: 3.5.2
```

</details>

--- a/py-polars/polars/testing/asserts.py
+++ b/py-polars/polars/testing/asserts.py
@@ -167,9 +167,14 @@ def _assert_series_inner(
         if left.dtype != right.dtype:
             raise_assert_detail(obj, "Dtype mismatch", left.dtype, right.dtype)
 
+    # confirm that we can call 'is_nan' on both sides
+    left_is_float = left.dtype in (Float32, Float64)
+    right_is_float = right.dtype in (Float32, Float64)
+    comparing_float_dtypes = left_is_float and right_is_float
+
     # create mask of which (if any) values are unequal
     unequal = left != right
-    if unequal.any() and nans_compare_equal and left.dtype in (Float32, Float64):
+    if unequal.any() and nans_compare_equal and comparing_float_dtypes:
         # handle NaN values (which compare unequal to themselves)
         unequal = unequal & ~(
             (left.is_nan() & right.is_nan()).fill_null(pli.lit(False))
@@ -182,13 +187,25 @@ def _assert_series_inner(
                 obj, "Exact value mismatch", left=list(left), right=list(right)
             )
         else:
-            # apply check with tolerance, but only to the known-unequal matches
+            # apply check with tolerance (to the known-unequal matches).
             left, right = left.filter(unequal), right.filter(unequal)
+            mismatch, nan_info = False, ""
             if (((left - right).abs() > (atol + rtol * right.abs())).sum() != 0) or (
-                (left.is_null() != right.is_null()).any()
-            ):
+                left.is_null() != right.is_null()
+            ).any():
+                mismatch = True
+            elif comparing_float_dtypes:
+                # note: take special care with NaN values.
+                if not nans_compare_equal and (left.is_nan() == right.is_nan()).any():
+                    nan_info = " (nans_compare_equal=False)"
+                    mismatch = True
+                elif (left.is_nan() != right.is_nan()).any():
+                    nan_info = f" (nans_compare_equal={nans_compare_equal})"
+                    mismatch = True
+
+            if mismatch:
                 raise_assert_detail(
-                    obj, "Value mismatch", left=list(left), right=list(right)
+                    obj, f"Value mismatch{nan_info}", left=list(left), right=list(right)
                 )
 
 

--- a/py-polars/tests/unit/test_testing.py
+++ b/py-polars/tests/unit/test_testing.py
@@ -20,17 +20,49 @@ def test_compare_series_empty_equal() -> None:
 
 
 def test_compare_series_nans_assert_equal() -> None:
-    # NaN values do not _compare_ equal, but should _assert_ as equal here
+    # note: NaN values do not _compare_ equal, but should _assert_ equal (by default)
     nan = float("NaN")
 
-    srs1 = pl.Series([1.0, 2.0, nan])
-    srs2 = pl.Series([1.0, 2.0, nan])
-    assert_series_equal(srs1, srs2)
+    srs1 = pl.Series([1.0, 2.0, nan, 4.0, None, 6.0])
+    srs2 = pl.Series([1.0, nan, 3.0, 4.0, None, 6.0])
+    srs3 = pl.Series([1.0, 2.0, 3.0, 4.0, None, 6.0])
 
-    srs1 = pl.Series([1.0, 2.0, nan])
-    srs2 = pl.Series([1.0, nan, 3.0])
+    for srs in (srs1, srs2, srs3):
+        assert_series_equal(srs, srs)
+        assert_series_equal(srs, srs, check_exact=True)
+
     with pytest.raises(AssertionError):
-        assert_series_equal(srs1, srs2, check_exact=True)
+        assert_series_equal(srs1, srs1, nans_compare_equal=False)
+    with pytest.raises(AssertionError):
+        assert_series_equal(srs1, srs1, nans_compare_equal=False, check_exact=True)
+
+    for check_exact, nans_equal in (
+        (False, False),
+        (False, True),
+        (True, False),
+        (True, True),
+    ):
+        if check_exact:
+            check_msg = "Exact value mismatch"
+        else:
+            check_msg = f"Value mismatch.*nans_compare_equal={nans_equal}"
+
+        with pytest.raises(AssertionError, match=check_msg):
+            assert_series_equal(
+                srs1, srs2, check_exact=check_exact, nans_compare_equal=nans_equal
+            )
+        with pytest.raises(AssertionError, match=check_msg):
+            assert_series_equal(
+                srs1, srs3, check_exact=check_exact, nans_compare_equal=nans_equal
+            )
+
+    srs4 = pl.Series([1.0, 2.0, 3.0, 4.0, None, 6.0])
+    srs5 = pl.Series([1.0, 2.0, 3.0, 4.0, nan, 6.0])
+    srs6 = pl.Series([1, 2, 3, 4, None, 6])
+
+    assert_series_equal(srs4, srs6, check_dtype=False)
+    with pytest.raises(AssertionError):
+        assert_series_equal(srs5, srs6, check_dtype=False)
 
 
 def test_compare_series_nulls() -> None:

